### PR TITLE
Cart feature

### PR DESCRIPTION
**Pull request by DaniyalXprolabs**

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,3 +1,30 @@
 class ApplicationController < ActionController::Base
   include DeviseWhitelist
+
+  def getCart
+    @cart = current_user.cart
+  end
+
+  def current_buyer
+    if session[:user_id]
+      @user=User.find(session[:user_id])
```
Comment (zunairzexprolabs) [inline app/controllers/application_controller.rb:10]: indentation

there should be space around =

File: app/controllers/cart_products_controller.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
+class CartProductsController < ApplicationController
+  before_action :getProduct, only: [:new]
+  before_action :getCart, only: [:new,:destroy,:add_quantity]
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:3]: space after ,

File: app/controllers/cart_products_controller.rb [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+  def new
+    if @cart.products.include?(@product)
+      @cart_product = @cart.cart_products.find_by(:product_id => @product.id)
+      @cart_product.quantity +=1
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:8]: space

File: app/controllers/cart_products_controller.rb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+      @cart_product.product_id = @product.id
+    end
+    @cart_product.subtotal = @product.price * @cart_product.quantity
+    @cart_product.save
+    redirect_to product_path(@product), notice: "Product add to the cart Successfully"
+  end
+
+  def add_quantity
+    @cart_product = CartProduct.find(params[:id])
+    @cart_product.quantity +=1
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:20]: same

File: app/controllers/cart_products_controller.rb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    if @cart.products.include?(@product)
+      @cart_product = @cart.cart_products.find_by(:product_id => @product.id)
+      @cart_product.quantity +=1
+    else
+      @cart_product = @cart.cart_products.new
+      @cart_product.product_id = @product.id
+    end
+    @cart_product.subtotal = @product.price * @cart_product.quantity
+    @cart_product.save
+    redirect_to product_path(@product), notice: "Product add to the cart Successfully"
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:15]: why we're not using respond_to here?

File: app/controllers/cart_products_controller.rb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    @cart_product = CartProduct.find(params[:id])
+    @cart_product.quantity +=1
+    @cart_product.subtotal = @cart_product.product.price * @cart_product.quantity
+    @cart_product.save
+    redirect_to cart_path(@cart), notice: "Product add to the cart Successfully"
+  end
+
+  def remove_quantity
+    @cart_product = CartProduct.find(params[:id])
+    @cart_product.quantity -=1
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:28]: space

File: app/controllers/cart_products_controller.rb [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  def add_quantity
+    @cart_product = CartProduct.find(params[:id])
+    @cart_product.quantity +=1
+    @cart_product.subtotal = @cart_product.product.price * @cart_product.quantity
+    @cart_product.save
+    redirect_to cart_path(@cart), notice: "Product add to the cart Successfully"
+  end
+
+  def remove_quantity
+    @cart_product = CartProduct.find(params[:id])
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:27]: add filter for this? shouldn't we

File: app/controllers/cart_products_controller.rb [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+    @cart_product.quantity -=1
+    @cart_product.subtotal = @cart_product.product.price * @cart_product.quantity
+    @cart_product.save
+    redirect_to cart_path(@cart), notice: "Product remove from cart Successfully"
+  end
+
+  def destroy
+    @cart_product = @cart.cart_products.find(params[:id])
```
Comment (zunairzexprolabs) [inline app/controllers/cart_products_controller.rb:35]: filter

File: app/controllers/carts_controller.rb [hunk 9]
```diff
@@ -0,0 +1,11 @@
+class CartsController < ApplicationController
+  before_action :getCart
```
Comment (zunairzexprolabs) [inline app/controllers/carts_controller.rb:2]: leave a line after it

File: app/controllers/products_controller.rb [hunk 10]
```diff
@@ -1,6 +1,7 @@
 class ProductsController < ApplicationController
   before_action :getProductId, only:[:show, :edit, :update,:destroy]
```
Comment (zunairzexprolabs) [inline app/controllers/products_controller.rb:2]: space

File: app/helpers/cart_products_helper.rb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+module CartProductsHelper
```
Comment (zunairzexprolabs) [inline app/helpers/cart_products_helper.rb:1]: why we included this file?

File: app/helpers/carts_helper.rb [hunk 12]
```diff
@@ -0,0 +1,2 @@
+module CartsHelper
```
Comment (zunairzexprolabs) [inline app/helpers/carts_helper.rb:1]: why?

File: app/models/cart.rb [hunk 13]
```diff
@@ -0,0 +1,5 @@
+class Cart < ApplicationRecord
+  belongs_to :user
+  has_many :cart_products
```
Comment (zunairzexprolabs) [inline app/models/cart.rb:3]: don't we need dependent destroy?

File: app/models/product.rb [hunk 14]
```diff
@@ -1,5 +1,8 @@
 class Product < ApplicationRecord
   belongs_to :user
+
```
Comment (zunairzexprolabs) [inline app/models/product.rb:3]: why this line?

File: app/models/user.rb [hunk 15]
```diff
@@ -1,4 +1,6 @@
 class User < ApplicationRecord
+  after_create :create_user_cart
```
Comment (zunairzexprolabs) [inline app/models/user.rb:2]: leave a line

File: app/controllers/cart_products_controller.rb (hunk 5 above)
Comment (DaniyalXprolabs) [reply app/controllers/cart_products_controller.rb:15]: Why do we have to use respond_to instead of redirect_to?

**Pull request merged**